### ISSUE TITLE
array is broken?

### ISSUE THREAD

**Issue by envirocoder** [skulpt/skulpt#1229]
I'm trying to index into an array and I get 'too much recursion error'.

I've checked it with the repl and demo on https://skulpt.org/ and it's the same behaviour as my instance:

```
from array import array
x = array('L',(1,2))
x[0]
ExternalError: InternalError: too much recursion on line 1
1>: evaluationresult = x[0]
2 : if not evaluationresult == None: print repr(evaluationresult)
```
Am I doing something wrong or is array broken?

I had a look in the source, but couldn't figure out how it was working -- `var functions` never seemed to be used!

**Comment by s-cork** [comment]
I have a feeling that `array` is recursively calling its own `__getattribute__`

https://github.com/skulpt/skulpt/blob/b763d3e24860a8e3d303878e0035d8bad9f55578/src/lib/array.js#L131

so I think you're not doing anything wrong and `array` is broken.
Array isn't optimised in skulpt - so using list might be better here.

**Comment by envirocoder** [comment]
Yes, I suspected the same, took out `__getattribute__` and added a `__getitem__`  and got something that works well enough for my needs.
I haven't got my head round the differences between `mp$subscript`, `__getitem__`  (and possibly others). Also don't know whether to use javascript's prototypal mechanism or though `buildClass` -- different examples seem to be implemented in different ways. 
Happy to try and help and submit a patch if there's any pointers to docs or pointers to a specific example of the "right way" to do things

**Comment by s-cork** [comment]
if you're defining something with `buildClass` then you should use `dunders`.
but you'll see most of the builtins don't use `buildClass` and that's for efficiency. 

for every python `dunder` function there's an equivalent raw javascript function.

If the aim is to patch `array` then i'd stick with `dunders` and the `buildClass` architecture`
If the aim is to build a more efficient native implementation of `array` then i'd look to use the raw javascript functions.

I'd start small - you could quite easily find yourself down a python/skulpt hole.
`re.js` has some more examples of `buildClass` as does `webbroswer.js`, `turtle.js`, `document.js` and `image.js`

There aren't any tests in skulpt for array - but any patch should come with some tests. It's best to get these from cpython
and it's fine to comment out tests that don't currently pass or are not appropriate for skulpt
https://github.com/python/cpython/blob/master/Lib/test/test_array.py
